merkle-tree/src/oneshot.rs

```diff
@@ -5,19 +5,73 @@ use simperby_common::MerkleProof;
 ///
 /// This is useful for per-block data such as transaction lists.
 pub struct OneshotMerkleTree {
-    // TODO
+    hash_list: Vec<Hash256>,
 }
 
 impl OneshotMerkleTree {
-    pub fn create(_data: Vec<Hash256>) -> Self {
-        unimplemented!()
+    pub fn create(data: Vec<Hash256>) -> Self {
+        OneshotMerkleTree { hash_list: data }
     }
 
-    pub fn create_merkle_proof(&self, _key: Hash256) -> MerkleProof {
-        unimplemented!()
+    pub fn create_merkle_proof(&self, key: Hash256) -> MerkleProof {
+        let proof: MerkleProof = MerkleProof { proof: Vec::new() };
+
+        Self::merkle_proof(&self.hash_list, key, proof)
+    }
+
+    fn merkle_proof(
+        hash_list: &Vec<Hash256>,
+        mut key: Hash256,
+        mut proof: MerkleProof,
+    ) -> MerkleProof {
+        if hash_list.len() == 1 {
+            return proof;
+        }
+
+        let mut upper_level_hash_list: Vec<Hash256> = Vec::new();
+
+        for pair in hash_list.chunks(2) {
+            let upper_level_hash: Hash256 = Self::hash_pair(pair);
+
+            if pair.contains(&key) {
+                let index: usize = (pair[0] == key) as usize;
+                proof.proof.push(pair[index].clone());
+                key = upper_level_hash.clone();
+            }
+
+            upper_level_hash_list.push(upper_level_hash);
+        }
+
+        Self::merkle_proof(&upper_level_hash_list, key, proof)
+    }
+
+    fn hash_pair(pair: &[Hash256]) -> Hash256 {
+        let first_hash: [u8; 32] = pair[0].hash;
+
+        let second_hash: [u8; 32] = if pair.len() == 2 {
+            pair[1].hash
+        } else {
+            first_hash
+        };
+
+        Hash256::hash([first_hash, second_hash].concat())
     }
 
     pub fn root(&self) -> Hash256 {
-        unimplemented!()
+        Self::hash_root(&self.hash_list)
+    }
+
+    fn hash_root(hash_list: &Vec<Hash256>) -> Hash256 {
```

Comment:
If you were to make `merkle_proof` return a list then it might be also generalized to implement this function.

Reply:
Applied